cores/arduino/stm32/usb/usbd_if.c

```diff
@@ -10,33 +10,77 @@
+// driven low very briefly, which looks like a missing pullup to the
+// host. Probably violates USB specs, but works properly in practice.
+#if !defined(USBD_ATTACH_PINNAME) && !defined(USBD_DETACH_PINNAME) && !defined(USB_OTG_DCTL_SDIS)
+#if defined(USE_USB_HS_IN_FS)
+#define USBD_DETACH_PINNAME USB_OTG_HS_DP
+#elif defined(USB_OTG_FS)
+#define USBD_DETACH_PINNAME USB_OTG_FS_DP
+#else /* USB */
+#define USBD_DETACH_PINNAME  USB_DP
+#endif
+#define USBD_DETACH_LEVEL LOW
+// Make sure to revert to INPUT (instead of OUTPUT HIGH) after the
+// detach "pulse" to restore normal operation.
+#define USBD_ATTACH_MODE INPUT
+#endif /* !defined(USBD_ATTACH_PINNAME) && !defined(USBD_DETACH_PINNAME) && !defined(USB_OTG_DCTL_SDIS) */
+
+#if defined(USBD_DETACH_PINNAME) && !defined(USBD_ATTACH_MODE)
+// Normally (e.g. when controlling a transistor), keep the pin in
+// OUTPUT mode after the detach "pulse" to actively drive the
+// transistor.
+#define USBD_ATTACH_MODE OUTPUT
+#endif /* defined(USBD_ATTACH_PINNAME) && !defined(USBD_ATTACH_MODE) */
+
+// Some sanity checks
+#if defined(USBD_ATTACH_PINNAME) && defined(USBD_DETACH_PINNAME)
+#error "Cannot define both USBD_ATTACH_PINNAME and USBD_DETACH_PINNAME"
+#endif /* defined(USBD_ATTACH_PINNAME) && defined(USBD_DETACH_PINNAME) */
+#if defined(USBD_ATTACH_PINNAME) && !defined(USBD_ATTACH_LEVEL)
+#error "USBD_ATTACH_PINNAME also needs USBD_ATTACH_LEVEL defined"
+#endif /* defined(USBD_ATTACH_PINNAME) && !defined(USBD_ATTACH_LEVEL) */
+#if defined(USBD_DETACH_PINNAME) && !defined(USBD_DETACH_LEVEL)
+#error "USBD_DETACH_PINNAME also needs USBD_DETACH_LEVEL defined"
+#endif /* defined(USBD_DETACH_PINNAME) && !defined(USBD_DETACH_LEVEL) */
+
 /**
   * @brief  Force to re-enumerate USB
   * @param  None
   * @retval None
   */
-
 WEAK void USBD_reenumerate(void)
 {
 #ifndef USBD_REENUM_DISABLED
   /* Re-enumerate the USB */
-#ifdef USB_DISC_PIN
-  pinMode(USB_DISC_PIN, OUTPUT);
-  digitalWrite(USB_DISC_PIN, LOW);
-#else
-#ifdef USE_USB_HS_IN_FS
-  PinName pinDP = USB_OTG_HS_DP;
-#elif defined(USB_OTG_FS)
-  PinName pinDP = USB_OTG_FS_DP;
-#else /* USB */
-  PinName pinDP =  USB_DP;
-#endif
-  pin_function(pinDP, STM_PIN_DATA(STM_MODE_OUTPUT_PP, GPIO_NOPULL, 0));
-  digitalWriteFast(pinDP, LOW);
+#if defined(USBD_DETACH_PINNAME)
+  // Default is attached, so detach, wait, and reattach.
+  pin_function(USBD_DETACH_PINNAME, STM_PIN_DATA(STM_MODE_OUTPUT_PP, GPIO_NOPULL, 0));
+  digitalWriteFast(USBD_DETACH_PINNAME, USBD_DETACH_LEVEL);
   delay(USBD_ENUM_DELAY);
-  pin_function(pinDP, STM_PIN_DATA(STM_MODE_INPUT, GPIO_NOPULL, 0));
-  /*delay(USBD_ENUM_DELAY);*/
-#endif /* USB_DISC_PIN */
+  if (USBD_ATTACH_MODE == OUTPUT) {
+    digitalWriteFast(USBD_DETACH_PINNAME, !USBD_DETACH_LEVEL);
+  } else {
+    pin_function(USBD_DETACH_PINNAME, STM_PIN_DATA(STM_MODE_INPUT, GPIO_NOPULL, 0));
+  }
+#elif defined(USBD_ATTACH_PINNAME)
+  // Default is detached, so just attach
```

Comment:
Reading your comment,

> Looking at the above, it seems that the first three cases are pretty identical (**the third one needs some extra work in case no reset happened beforehand, but that extra work should not hurt if a reset happened, so maybe best to just always do that**).

I was expecting to see similar processing for both USBD_ATTACH_PINNAME and USBD_DETACH_PINNAME : 
First dettach then attach, thus you would cover the third case.

Reply:
Yeah, let's do that indeed.